libcache/cache.c

```diff
@@ -17,25 +17,54 @@
 
 cacheline_t cache_createLine(uint64_t tag, uint32_t *data)
 {
-	cacheline_t line = {tag, data, '1'};
+	cacheline_t line = { tag, data, '1' };
 
-	return line;	
+	return line;
 }
 
 cacheset_t *cache_createSet(void)
 {
-	cacheset_t *cache_set = malloc(sizeof(cacheset_t)); 
+	int i = 0;
+	cacheset_t *cacheSet = malloc(sizeof(cacheset_t));
 
-	if (cache_set == NULL)
+	if (cacheSet == NULL) {
 		return NULL;
-	
-	cacheline_t cache_line = {0, 0, '0'};
+	}
+
+	cacheSet->timestamps = calloc(NUM_WAYS, sizeof(cacheline_t *));
+	cacheSet->tags = calloc(NUM_WAYS, sizeof(cacheline_t *));
+	cacheSet->lines = calloc(NUM_WAYS, sizeof(cacheline_t));
 
-	for (int i = 0; i < NUM_WAYS; ++i) {
-		cache_set->timestamps[i] = NULL;
-		cache_set->tags[i] = NULL;
-		cache_set->lines[i] = cache_line; 	
+	if (cacheSet->timestamps == NULL || cacheSet->tags == NULL || cacheSet->lines == NULL) {
+		cache_freeSet(cacheSet);
+		return NULL;
 	}
 
-	return cache_set;
+	return cacheSet;
 }
+
+int cache_destroySet(cacheset_t *cacheSet)
+{
+	if (cacheSet == NULL) {
+		return -1;
+	}
+
+	cache_freeSet(cacheSet);
+
+	return 0;
+}
+
+void cache_freeSet(cacheset_t *cacheSet)
+{
+	if (cacheSet->timestamps != NULL) {
```

Comment:
spotted in e-mail - `free(NULL)` is a no-op so these `if`'s can be safely deleted ;). Also - my personal preference is to reverse the order of allocations while freeing (might help stupid allocator implementations).

Reply:
Thank you for pointing this out. I'll fix it.